### PR TITLE
[node-addon-api] Add new port

### PR DESCRIPTION
**Pull request by Pospelove**
depends on https://github.com/microsoft/vcpkg/pull/26768

node-api is official C API for NodeJS

node-addon-api is C++ wrapper around node-api

### CODE REVIEW

**Comment by JonLiu1993** [comment]
@Pospelove, Thanks for your pr, if this pr is ready for review please let me know.

**Comment by Pospelove** [comment]
@JonLiu1993 thank you, it's ready

however merging this would depend on https://github.com/microsoft/vcpkg/pull/26768

**Comment by Pospelove** [comment]
@JonLiu1993 ready

File: ports/node-addon-api/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,14 @@
+vcpkg_from_github(
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO nodejs/node-addon-api
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
```
Comment (JonLiu1993) [inline ports/node-addon-api/portfile.cmake:5]:
```suggestion
  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
HEAD_REF main
```

File: ports/node-addon-api/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+vcpkg_from_github(
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO nodejs/node-addon-api
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+)
+
+file(COPY ${SOURCE_PATH}/napi.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.deprecated.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
```
Comment (JonLiu1993) [inline ports/node-addon-api/portfile.cmake:10]:
```suggestion
file(COPY "${SOURCE_PATH}/napi.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
file(COPY "${SOURCE_PATH}/napi-inl.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
file(COPY "${SOURCE_PATH}/napi-inl.deprecated.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
```

File: ports/node-addon-api/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+)
+
+file(COPY ${SOURCE_PATH}/napi.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.deprecated.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+
+file(INSTALL ${SOURCE_PATH}/LICENSE.md DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
+
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/unofficial-node-addon-api-config.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT})
```
Comment (JonLiu1993) [inline ports/node-addon-api/portfile.cmake:14]:
```suggestion
file(COPY "${CMAKE_CURRENT_LIST_DIR}/unofficial-node-addon-api-config.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```

File: ports/node-addon-api/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "node-addon-api",
+  "version-semver": "5.0.0",
```
Comment (JonLiu1993) [inline ports/node-addon-api/vcpkg.json:3]:
```suggestion
  "version": "5.0.0",
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: ports/node-addon-api/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,15 @@
+vcpkg_from_github(
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO nodejs/node-addon-api
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+HEAD_REF main
```
Comment (JonLiu1993) [inline ports/node-addon-api/portfile.cmake:6]:
```suggestion
  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
  HEAD_REF main
```

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by JonLiu1993** [comment]
@Pospelove , Thanks for your pr, when I tested the usage I got this error, please take a look:

CMakeLists.txt:
````
# CMakeList.txt : CMake project for CMakeProject42, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

project(test)

find_package(node-addon-api CONFIG REQUIRED)

add_executable (main "CMakeProject42.cpp" "CMakeProject42.h")

target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)

````


CMakeProject42.cpp:
````
// CMakeProject42.cpp : Defines the entry point for the application.
//

#include "CMakeProject42.h"
#include "napi.h"

using namespace std;

int main()
{
	cout << "Hello CMake." << endl;
	return 0;
}

````
CMakeProject42.h:
````
// CMakeProject42.h : Include file for standard system include files,
// or project specific include files.

#pragma once

#include <iostream>

// TODO: Reference additional headers your program requires here.

````

````
1> CMake generation started for configuration: 'x64-Debug'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\vzhli17\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\vzhli17\source\repos\CMakeProject42" 2>&1"
1> Working directory: C:\Users\test\source\repos\CMakeProject42\out\build\x64-Debug
1> [CMake] -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
1> [CMake] -- The C compiler identification is MSVC 19.29.30146.0
1> [CMake] -- The CXX compiler identification is MSVC 19.29.30146.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] CMake Error at F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake:843 (_find_package):
1> [CMake]   Could not find a package configuration file provided by "node-addon-api"
1> [CMake]   with any of the following names:
1> [CMake] 
1> [CMake]     node-addon-apiConfig.cmake
1> [CMake]     node-addon-api-config.cmake
1> [CMake] 
1> [CMake]   Add the installation prefix of "node-addon-api" to CMAKE_PREFIX_PATH or set
1> [CMake]   "node-addon-api_DIR" to a directory containing one of the above files.  If
1> [CMake]   "node-addon-api" provides a separate development package or SDK, be sure it
1> [CMake]   has been installed.
1> [CMake] Call Stack (most recent call first):
1> [CMake]   CMakeProject42/CMakeLists.txt:10 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
1> [CMake] See also "C:/Users/test/source/repos/CMakeProject42/out/build/x64-Debug/CMakeFiles/CMakeOutput.log".
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\test\source\repos\CMakeProject42" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\test\source\repos\CMakeProject42" 2>&1"' returned with exit code: 1'.


````

**Comment by dg0yt** [comment]
> find_package(node-addon-api CONFIG REQUIRED)
> target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)

Shouldn't this be
~~~cmake
find_package(unofficial-node-addon-api CONFIG REQUIRED)
~~~
or similar?

**Comment by Pospelove** [comment]
wow. it seems both are not working for me for some reason. 

unofficial-node-addon-api and node-addon-api 

investigating

**Comment by Pospelove** [comment]
I see. Fixing.

File: ports/node-addon-api/unofficial-node-addon-api-config.cmake [hunk 6]
```diff
@@ -0,0 +1,13 @@
+find_package(unofficial-node-api CONFIG REQUIRED)
```
Comment (dg0yt) [inline ports/node-addon-api/unofficial-node-addon-api-config.cmake:1]: This should be 
~~~cmake
include(CMakeFindDependencyMacro)
find_dependency(unofficial-node-api CONFIG)
~~~

File: ports/node-addon-api/unofficial-node-addon-api-config.cmake [hunk 7]
```diff
@@ -0,0 +1,13 @@
+find_package(unofficial-node-api CONFIG REQUIRED)
+
+add_library(unofficial::node-addon-api::node-addon-api INTERFACE)
+target_link_libraries(unofficial::node-addon-api::node-addon-api INTERFACE unofficial::node-api::node-api)
+
+find_path(node-addon-api_INCLUDE_DIR
+  NAMES napi.h
+  PATHS "${CMAKE_CURRENT_LIST_DIR}/../../include/node-addon-api"
+  NO_DEFAULT_PATH
+  REQUIRED)
```
Comment (dg0yt) [inline ports/node-addon-api/unofficial-node-addon-api-config.cmake]: I don't think `find_path` is needed here: The path is already known. Just insert into `INTERFACE_INCLUDE_DIRECTORIES`.

File: ports/node-addon-api/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+  HEAD_REF main
+)
+
+file(COPY "${SOURCE_PATH}/napi.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+file(COPY "${SOURCE_PATH}/napi-inl.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+file(COPY "${SOURCE_PATH}/napi-inl.deprecated.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+
+file(INSTALL ${SOURCE_PATH}/LICENSE.md DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Comment (dg0yt) [inline ports/node-addon-api/portfile.cmake:13]: Missing quotes.
I would prefer if this was the last line in the portfile.

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by Pospelove** [comment]
@JonLiu1993 Works now. Changed your cmakelists a bit:

added cxx standard and "unofficial" prefix for find_package

```cmake
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_LIST_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake")

# CMakeList.txt : CMake project for CMakeProject42, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

set(CMAKE_CXX_STANDARD 20) # !!! I added this 

project(test)

find_package(unofficial-node-addon-api CONFIG REQUIRED) # !!! and added unofficial word here

add_executable (main "CMakeProject42.cpp" "CMakeProject42.h")

target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)
```

@dg0yt thanks. taking a look

**Comment by Pospelove** [comment]
@dg0yt  @JonLiu1993  works and fixed

**Comment by Pospelove** [comment]
@JonLiu1993

**Comment by JonLiu1993** [comment]
> @JonLiu1993

@Pospelove, sorry for reply you so late, after the CI turns green, I will put the reviewed label on the table after the local test is successful.

**Comment by Pospelove** [comment]
@JonLiu1993
plz consider taking a look

**Comment by Pospelove** [comment]
@JonLiu1993

**Comment by JonLiu1993** [comment]
@Pospelove, sorry to reply you so late, we just came back from vacation, so we didn't respond to you in time.

**Comment by JonLiu1993** [comment]
Tested usage by `node-addon-api:x64-windows` triplet successfully.

**Review by BillyONeal — APPROVED** [review]
Thanks for the new port!

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.  
  Note: There is a 5.1.0 already :)
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See docs/examples/adding-an-explicit-usage.md for context.  
  I added usage text
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

**Pull request merged**